include/arch/arm/arch/64/mode/hardware.h

```diff
@@ -218,9 +219,13 @@
    overflow without going into address ranges that are non-canonical.  These static
    asserts check that the kernel config won't lead to UTs being created that aren't
    representable. */
-#ifndef __ASSEMBLER__
 compile_assert(ut_max_less_than_cannonical, CONFIG_PADDR_USER_DEVICE_TOP <= BIT(47));
 #ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
 compile_assert(ut_max_is_cannonical, (PPTR_BASE + CONFIG_PADDR_USER_DEVICE_TOP) <= BIT(48));
 #endif
+
+/* PPTR_BASE must be smaller than KERNEL_ELFBASE. KERNEL_ELF_BASE is unfolded below
+   because the compiler refuses to do so automatically. */
```

Comment:
Maybe "refuse" is too strong. `KERNEL_ELF_BASE` is built using `KERNEL_ELF_PADDR_BASE`, which isince 65a1b457a4ba3e72838434a68c5ce9889d3ce406 is the result of a function call `physBase()`. While this is a inlined const function, sadly in C such things are still not evaluated at compile time to be used in expressions then (zig can do this, maybe there is a gcc/clang switch also?). What we could do here is using `KERNEL_ELF_BASE_RAW` and state it the comment, why this is the better choice..

Reply:
Seems with c23 we could do change https://github.com/seL4/seL4/blob/master/tools/hardware/outputs/c_header.py#L43-L46 to use `constexpr word_t physBase = PHYS_BASE_RAW;`. Unfortunately `constexpr` for function is still not in the standard.

Reply:
I tried using `KERNEL_ELF_BASE_RAW` initially, but it was complaining about `PPTR_BASE_OFFSET` which is used in both. It's possible that the definition of `PPTR_BASE_OFFSET` just wasn't included yet at this point. Maybe I should have another look at that.

I don't think we want to switch to C23 yet at this point, although some features might be nice to cherry pick. In any case, not for this PR, though.

Reply:
Ok, yes including `<hardware.h>` does fix the problem and `KERNEL_ELF_BASE_RAW` then works (see additional commit). The generic `<hardware.h>` is the place where `PPTR_BASE_OFFSET` is defined. It is a bit strange to include it here, although it does appear to compile fine.

I'd be fine either way (unfolding `KERNEL_ELF_BASE_RAW` and `PPTR_BASE_OFFSET` here or including generic  `<hardware.h>` at the end).  @kent-mcleod do you have any preferences for this?

Reply:
The point has become irrelevant, because the check is not actually necessary for AArch64 (at least for the verified hyp config), so I'll remove the check and mark this as resolved.